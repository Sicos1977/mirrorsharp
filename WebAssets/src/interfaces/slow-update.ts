import type { DiagnosticData } from './protocol';

export interface SlowUpdateOptions<TSlowUpdateExtensionData> {
    slowUpdateWait?: () => void;
    slowUpdateResult?: (args: { diagnostics: ReadonlyArray<DiagnosticData>; x: TSlowUpdateExtensionData }) => void;
}