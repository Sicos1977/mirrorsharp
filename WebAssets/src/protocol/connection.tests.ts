import { TestDriver } from '../testing/test-driver-jest';

test('attempts to reopen connection if lost', async () => {
    const driver = await TestDriver.new({});

    driver.socket.close();
    await driver.completeBackgroundWork();

    expect(driver.socket.createdCount).toBe(2);
});