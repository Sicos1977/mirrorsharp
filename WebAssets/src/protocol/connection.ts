import { ensureDefined } from '../helpers/ensure-defined';
import type { Message, ServerOptions } from './messages';

const stateCommandMap = { cancel: 'X', force: 'F' } as Readonly<{
    cancel: 'X';
    force: 'F';
    [key: number]: undefined;
}>;

export type ReplaceTextCommand = {
    start: number,
    length: number,
    newText: string,
    cursorIndexAfter: number,
    reason?: string | null
};

type HandlerMap<O, U> = {
    open:    () => void,
    message: (data: Message<O, U>) => void,
    close:   () => void
};

// Defaults are 'unknown' rather than 'void', as it exists for internal convenience,
// and we assume in most cases this is not 'void'. Anything public should have 'void' though.
export class Connection<TExtensionServerOptions = unknown, TSlowUpdateExtensionData = unknown> {
    readonly #url: string;

    readonly #listeners = {
        open:    [],
        message: [],
        close:   []
    } as {
        [K in keyof HandlerMap<TExtensionServerOptions, TSlowUpdateExtensionData>]:
            Array<HandlerMap<TExtensionServerOptions, TSlowUpdateExtensionData>[K]>
    };

    #socket: WebSocket | undefined;
    #openPromise!: Promise<void>;
    #resolveDelayedOpenPromise: (() => void) | undefined | null;

    #mustBeClosed = false;
    #reopenPeriod = 0;
    #reopenPeriodResetTimer: ReturnType<typeof setTimeout> | undefined | null;
    #reopening = false;

    #removeInternalListeners: () => void;

    constructor(url: string, { closed }: { closed: boolean | undefined }) {
        this.#url = url;

        if (!closed) {
            this.open();
        }
        else {
            this.#openPromise = new Promise(resolve => {
                this.#resolveDelayedOpenPromise = resolve;
            });
        }

        this.#removeInternalListeners = this.addEventListeners({
            close: () => this.#tryToReopen()
        });
    }

    addEventListeners(listeners: Partial<HandlerMap<TExtensionServerOptions, TSlowUpdateExtensionData>>) {
        const removeEach = [] as Array<() => void>;
        for (const key in listeners) {
            const list = this.#listeners[key as keyof typeof listeners];
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            const listener = listeners[key as keyof typeof listeners]!;

            list.push(
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                listener as any
            );
            removeEach.push(() => {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-call
                const index = list.indexOf(listener as any);
                if (index >= 0)
                    list.splice(index, 1);
            });
        }

        return () => {
            for (const remove of removeEach) {
                remove();
            }
        };
    }

    open() {
        this.#socket = new WebSocket(this.#url);
        this.#openPromise = new Promise(resolve => {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            this.#socket!.addEventListener('open', () => {
                this.#reopenPeriodResetTimer = setTimeout(() => { this.#reopenPeriod = 0; }, this.#reopenPeriod);
                if (this.#resolveDelayedOpenPromise) {
                    this.#resolveDelayedOpenPromise();
                    this.#resolveDelayedOpenPromise = null;
                }
                resolve();
            });
        });
        this.#addEventListenersToSocket();
    }

    #addEventListenersToSocket() {
        if (!this.#socket)
            throw new Error('Internal error: Socket was not created before adding listeners.');

        this.#socket.addEventListener('open', () => {
            for (const listener of this.#listeners['open']) {
                listener();
            }
        });
        this.#socket.addEventListener('message', e => {
            const data = JSON.parse(e.data as string) as Message<TExtensionServerOptions, TSlowUpdateExtensionData>;
            for (const listener of this.#listeners['message']) {
                listener(data);
            }
        });
        this.#socket.addEventListener('close', () => {
            for (const listener of this.#listeners['close']) {
                listener();
            }
        });
    }

    #tryToReopen() {
        if (this.#mustBeClosed || this.#reopening)
            return;

        if (this.#reopenPeriodResetTimer) {
            clearTimeout(this.#reopenPeriodResetTimer);
            this.#reopenPeriodResetTimer = null;
        }

        this.#reopening = true;
        setTimeout(() => {
            this.open();
            this.#reopening = false;
        }, this.#reopenPeriod);
        if (this.#reopenPeriod < 60000)
            this.#reopenPeriod = Math.min(5 * (this.#reopenPeriod + 200), 60000);
    }

    async #sendIfOpen(command: string) {
        if (this.#mustBeClosed)
            throw `Cannot send command '${command}' after the close() call.`;

        if (!this.isOpen()) {
            // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
            console.warn(`Dropped command '${command}' because the socket state is ${this.#socket?.readyState}.`);
            return;
        }

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        this.#socket!.send(command);
    }

    isOpen() {
        return this.#socket?.readyState === WebSocket.OPEN;
    }

    sendReplaceText({ start, length, cursorIndexAfter, newText, reason }: ReplaceTextCommand) {
        return this.#sendIfOpen('R' + start + ':' + length + ':' + cursorIndexAfter + ':' + (reason ?? '') + ':' + newText);
    }

    sendMoveCursor(cursorIndex: number) {
        return this.#sendIfOpen('M' + cursorIndex);
    }

    sendTypeChar(char: string) {
        return this.#sendIfOpen('C' + char);
    }

    sendCompletionState(indexOrCommand: 'info'|'force'|number|'cancel', indexIfInfo?: number) {
        // common bug -- the specific flow is not fully clear yet,
        // but there is no reason to send null/undefined to server if it will fail anyway
        ensureDefined(indexOrCommand, 'completion command');
        const argument = indexOrCommand !== 'info'
            ? (stateCommandMap[indexOrCommand] ?? indexOrCommand)
            : 'I' + ensureDefined(indexIfInfo, 'completion info index');
        return this.#sendIfOpen('S' + argument);
    }

    sendSignatureHelpState(command: 'force'|'cancel') {
        return this.#sendIfOpen('P' + stateCommandMap[command]);
    }

    sendRequestInfoTip(cursorIndex: number) {
        return this.#sendIfOpen('I' + cursorIndex);
    }

    sendSlowUpdate() {
        return this.#sendIfOpen('U');
    }

    sendApplyDiagnosticAction(actionId: number) {
        return this.#sendIfOpen('F' + actionId);
    }

    sendSetOptions(options: Partial<ServerOptions> & Partial<TExtensionServerOptions>) {
        const optionPairs = [];
        for (const key in options) {
            optionPairs.push(key + '=' + (options as Record<string, string>)[key]);
        }
        return this.#sendIfOpen('O' + optionPairs.join(','));
    }

    sendRequestSelfDebugData() {
        return this.#sendIfOpen('Y');
    }

    close() {
        this.#mustBeClosed = true;
        this.#removeInternalListeners();
        this.#socket?.close();
    }
}