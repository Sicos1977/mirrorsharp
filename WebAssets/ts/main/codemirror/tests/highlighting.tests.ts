import { LANGUAGE_FSHARP } from '../../../interfaces/protocol';
import { TestDriver } from '../../../testing/test-driver';

const CODE_CSHARP = `
public class C<T> {
    // test comment
    public void M<U>() {
        double d = 1.2e3;
        string s1 = "test";
        string s2 = $"a{s1}b";
        char c = 't';
        Action a = () => {};
    }
}`.replace(/\r\n|\r|\n/g, '\r\n').trim();

test('C# highlighting applies expected classes', async () => {
    const driver = await TestDriver.new({ text: CODE_CSHARP });
    await driver.completeBackgroundWork();

    const html = driver.getCodeMirrorView().contentDOM.innerHTML;

    expect(html).toMatchSnapshot();
});

test('C# highlighting is rendered correctly', async () => {
    if (TestDriver.shouldSkipRender)
        return;
    const driver = await TestDriver.new({ text: CODE_CSHARP });
    await driver.completeBackgroundWork();

    const rendered = await driver.render();

    expect(rendered).toMatchImageSnapshot();
});

const CODE_FSHARP = `
type C<'a> =
    // test comment
    member this.M<'b>() =
        let d = 1.2e3
        let s1 = "test"
        let s2 = $"a{s1}b"
        let c = 't'
        let a = fun () -> ()
        ()
`.replace(/\r\n|\r|\n/g, '\r\n').trim();

test('F# highlighting applies expected classes', async () => {
    const driver = await TestDriver.new({
        text: CODE_FSHARP,
        options: { language: LANGUAGE_FSHARP }
    });
    await driver.completeBackgroundWork();

    const html = driver.getCodeMirrorView().contentDOM.innerHTML;

    expect(html).toMatchSnapshot();
});

test('F# highlighting is rendered correctly', async () => {
    if (TestDriver.shouldSkipRender)
        return;
    const driver = await TestDriver.new({
        text: CODE_FSHARP,
        options: { language: LANGUAGE_FSHARP }
    });
    await driver.completeBackgroundWork();

    const rendered = await driver.render();

    expect(rendered).toMatchImageSnapshot();
});