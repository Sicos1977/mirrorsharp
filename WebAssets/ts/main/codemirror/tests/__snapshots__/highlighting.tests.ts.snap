// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`C# highlighting applies expected classes 1`] = `"<div class="cm-line"><span class="tok-keyword">public</span> <span class="tok-keyword">class</span> C<span class="tok-punctuation">&lt;</span>T<span class="tok-punctuation">&gt;</span> <span class="tok-punctuation">{</span></div><div class="cm-line">    <span class="tok-comment">// test comment</span></div><div class="cm-line">    <span class="tok-keyword">public</span> <span class="tok-keyword">void</span> M<span class="tok-punctuation">&lt;</span>U<span class="tok-punctuation">&gt;</span><span class="tok-punctuation">(</span><span class="tok-punctuation">)</span> <span class="tok-punctuation">{</span></div><div class="cm-line">        <span class="tok-keyword">double</span> d <span class="tok-punctuation">=</span> <span class="tok-number">1.2e3</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-keyword">string</span> s1 <span class="tok-punctuation">=</span> <span class="tok-string">"test"</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-keyword">string</span> s2 <span class="tok-punctuation">=</span> $<span class="tok-string">"a{s1}b"</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-keyword">char</span> c <span class="tok-punctuation">=</span> 't'<span class="tok-punctuation">;</span></div><div class="cm-line">        Action a <span class="tok-punctuation">=</span> <span class="tok-punctuation">(</span><span class="tok-punctuation">)</span> <span class="tok-punctuation">=&gt;</span> <span class="tok-punctuation">{</span><span class="tok-punctuation">}</span><span class="tok-punctuation">;</span></div><div class="cm-line">    <span class="tok-punctuation">}</span></div><div class="cm-line"><span class="tok-punctuation">}</span></div>"`;

exports[`F# highlighting applies expected classes 1`] = `"<div class="cm-line"><span class="tok-keyword">type</span> <span class="tok-variableName">C</span><span class="tok-operator">&lt;</span>'<span class="tok-variableName">a</span><span class="tok-operator">&gt;</span> <span class="tok-operator">=</span></div><div class="cm-line">    <span class="tok-comment">// test comment</span></div><div class="cm-line">    <span class="tok-keyword">member</span> <span class="tok-variableName">this</span><span class="tok-operator">.</span><span class="tok-variableName">M</span><span class="tok-operator">&lt;</span>'<span class="tok-variableName">b</span><span class="tok-operator">&gt;</span>() <span class="tok-operator">=</span></div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">d</span> <span class="tok-operator">=</span> <span class="tok-number">1.2e3</span></div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">s1</span> <span class="tok-operator">=</span> <span class="tok-string">"test"</span></div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">s2</span> <span class="tok-operator">=</span> $<span class="tok-string">"a{s1}b"</span></div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">c</span> <span class="tok-operator">=</span> '<span class="tok-variableName">t</span>'</div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">a</span> <span class="tok-operator">=</span> <span class="tok-keyword">fun</span> () <span class="tok-operator">-</span><span class="tok-operator">&gt;</span> ()</div><div class="cm-line">        ()</div>"`;
